### PR TITLE
POC/WIP on provider package discovery mechanism

### LINKED ISSUE #11422: Automatically discover available packages
The packages installed should be automatically discoverable using Python mechanisms and a form of registration of those packages in the core should happen - similar to Plugin mechanism.

This is needed in order to expose various mechanisms where providers have to be registered in the code.

### PR DESCRIPTION
**Pull request by potiuk**
Here is a POC/WIP of dynamic discovery for Provider packages . 

This is a WIP of master version of the dynamic discovery of packages from #11163 . It needs completion. Details of the work needed are described in https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=100827303

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/get_provider_info.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    providers_path = providers.__path__
+    providers_name = providers.__name__
+    provider_dict = {}
+
+    for (_, name, ispkg) in pkgutil.walk_packages(path=providers_path,
+                                                  prefix=providers_name + ".",
+                                                  onerror=ignore):
+        try:
+            if ispkg:
```
Comment (turbaszek) [inline airflow/get_provider_info.py:64]:
```suggestion
                                                  onerror=ignore):
        if not ispkg:
            continue
        try:
```
WDYT?

File: airflow/get_provider_info.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import importlib
+import pkgutil
+
+
+def get_provider_info(conn_type_to_hook, connection_types):
```
Comment (ashb) [inline airflow/get_provider_info.py:22]: This feels similar to existing code in PluginsManager, and serves a similar purpose. What do you think about move this code in to there?

Comment (potiuk) [reply airflow/get_provider_info.py:22]: Sure - it's really a POC which I prepared quickly - happy to move it there.

**Comment by potiuk** [comment]
@kaxil - I missed this at the beginning of the call - this PR had "beta1" but I do not think it is actualy needed in beta1 (and this is a POC so I am going to re-implement it anyway in smaller pieces). I planned the work there for different betas to add those incrementally and split it into separate issues (all in https://github.com/apache/airflow/projects/5)

**Comment by ashb** [comment]
@potiuk I've been having a quick look at using https://pluggy.readthedocs.io/en/latest/ (the plugin system from pytest) for this -- I'll let you know my thoughts

**Comment by potiuk** [comment]
> @potiuk I've been having a quick look at using https://pluggy.readthedocs.io/en/latest/ (the plugin system from pytest) for this -- I'll let you know my thoughts

Interesting. I have not yet come back to it  - busy with CI optimisations, but will come back to it this week, so happy to hear your thoughts. this was just POC to see that it is possible but if there are ready-to-use solutions, I  am happy to use them instead

**Comment by ashb** [comment]
The main problem we we've run in to with the existing plugin mechanism (we ship some custom plugins in our Astronomer images) is that any conflicting deps means plugins are unloadable:

```
Traceback (most recent call last):
  File "/home/ash/code/airflow/airflow/airflow/plugins_manager.py", line 187, in load_entrypoint_plugins
    plugin_class = entry_point.load()
  File "/home/ash/.virtualenvs/airflow-with-plugins/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2442, in load
    self.require(*args, **kwargs)
  File "/home/ash/.virtualenvs/airflow-with-plugins/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2465, in require
    items = working_set.resolve(reqs, env, installer, extras=self.extras)
  File "/home/ash/.virtualenvs/airflow-with-plugins/lib/python3.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (lazy-object-proxy 1.2.2 (/home/ash/.virtualenvs/airflow-with-plugins/lib/python3.7/site-packages), Requirement.parse('lazy_object_proxy~=1.3'))
```

Given the odds of _some_ version conflict in a non-trival airflow install is quite high, using our existing plugin mechanism for more doesn't seem like a great idea.

I haven't looked in detail yet at Pluggy, but _somehow_ it doesn't suffer this same problem, even though it supports entrypoints.

Also at first though I quite like the idea of a richer plugin mechanism in Airflow.

**Pull request closed without merging**